playwright-e2e/functional/helpers/ApiBuilder.ts

```diff
@@ -0,0 +1,351 @@
+import axios, { AxiosRequestConfig, AxiosResponse } from "axios";
+import * as fs from "fs";
+import path from "path";
+import * as otp from "otplib";
+import { set, unset } from "lodash";
+
+const axiosClient = axios.create({});
+
+const env = process.env.RUNNING_ENV || "aat";
+const ccdApiUrl = process.env.CCD_DATA_API_URL;
+const idamBaseUrl = `https://idam-api.${env}.platform.hmcts.net`;
+
+interface ReplacementAction {
+  action: string;
+  key: string;
+  value?: any;
+}
+
+export class CcdApiHelper {
+  async axiosRequest<T = any>(
+    requestParams: AxiosRequestConfig
+  ): Promise<AxiosResponse<T> | undefined> {
+    try {
+      return await axiosClient(requestParams);
+    } catch (error: any) {
+      console.error(
+        "Utils %s request error %s",
+        requestParams.url,
+        error.message
+      );
+    }
+  }
+
+  async getUserToken(username: string, password: string): Promise<string> {
+    console.info("Getting User Token");
+
+    const idamClientSecret = process.env.IDAM_CLIENT_SECRET;
+    const redirectUri = `https://div-pfe-${env}.service.core-compute-${env}.internal/authenticated`;
+    const idamCodePath = `/oauth2/authorize?response_type=code&client_id=divorce&redirect_uri=${redirectUri}`;
+
+    const idamCodeResponse = await this.axiosRequest({
+      method: "post",
+      url: idamBaseUrl + idamCodePath,
+      headers: {
+        Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString("base64")}`,
+        "Content-Type": "application/x-www-form-urlencoded",
+      },
+    });
+
+    console.info("Successfully retrieved IdAM code");
+
+    const idamAuthPath = `/oauth2/token?grant_type=authorization_code&client_id=divorce&client_secret=${idamClientSecret}&redirect_uri=${redirectUri}&code=${idamCodeResponse?.data.code}`;
+
+    const authTokenResponse = await this.axiosRequest({
+      method: "post",
+      url: idamBaseUrl + idamAuthPath,
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded",
+      },
+    });
+
+    console.info("Successfully retrieved user token");
+    return authTokenResponse?.data.access_token;
+  }
```

Comment:
As this is new, consider looking at a valid redirect url.  This is Old Divorce's petitioner frontend.  Which we hope to turn off ASAP.